crates/bls12381/src/fields/fp.rs

```diff
@@ -0,0 +1,504 @@
+    // for example, in the pasta field, 4/96 could be used instead
+    fn num_limbs() -> usize {
+        6
+    }
+
+    fn bits_per_limb() -> usize {
+        64
+    }
+
+    fn modulus() -> BigInt {
+        BigInt::parse_bytes(
+            b"1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab",
+            16,
+        )
+        .unwrap()
+    }
+
+    fn is_modulus_pseudo_mersenne() -> bool {
+        false
+    }
+
+    fn pseudo_mersenne_params() -> Option<PseudoMersennePrime> {
+        None
+    }
+}
+
+pub type Bls12381Fp<F> = EmulatedFieldElement<F, Bls12381FpParams>;
+
+#[derive(Clone)]
+pub struct AllocatedFieldElement<F: PrimeField + PrimeFieldBits>(pub(crate) Bls12381Fp<F>);
+
+#[derive(Debug)]
+pub struct Bls12381FrParams;
+
+impl EmulatedFieldParams for Bls12381FrParams {
+    fn num_limbs() -> usize {
+        4
+    }
+
+    fn bits_per_limb() -> usize {
+        64
+    }
+
+    fn modulus() -> BigInt {
+        BigInt::parse_bytes(
+            b"73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001",
+            16,
+        )
+        .unwrap()
+    }
+
+    fn is_modulus_pseudo_mersenne() -> bool {
+        false
+    }
+
+    fn pseudo_mersenne_params() -> Option<PseudoMersennePrime> {
+        None
+    }
+}
+
+pub type Bls12381Fr<F> = EmulatedFieldElement<F, Bls12381FrParams>;
+
+impl<F> From<&BlsFp> for AllocatedFieldElement<F>
+where
+    F: PrimeField + PrimeFieldBits,
+{
+    fn from(value: &BlsFp) -> Self {
+        let bytes = value.to_bytes();
+        assert!(bytes.len() == 48);
+        let val = BigInt::from_bytes_be(Sign::Plus, &bytes);
+        Self(Bls12381Fp::<F>::from(&val))
+    }
+}
+
+pub(crate) fn bigint_to_fpelem(val: &BigInt) -> Option<BlsFp> {
+    use num_traits::Zero;
+    if val >= &Bls12381FpParams::modulus() {
+        return None;
+    }
+    let be_bytes = BigInt::to_bytes_be(val);
+    if be_bytes.0 != Sign::Plus {
+        // the sign is only non-positive if the value is exactly 0
+        assert!(val == &BigInt::zero());
+    }
+    let mut bytes: Vec<u8> = be_bytes.1;
+    assert!(bytes.len() <= 48);
+    while bytes.len() < 48 {
+        bytes.insert(0, 0);
+    } // TODO: here we pad with 0 bytes up to 48 bytes, but maybe something better could be done instead
+    let bytes: [u8; 48] = bytes.try_into().unwrap();
+    Some(BlsFp::from_bytes(&bytes).unwrap())
+}
+
+pub(crate) fn emulated_to_native<F>(value: &Bls12381Fp<F>) -> BlsFp
+where
+    F: PrimeField + PrimeFieldBits,
+{
+    use std::ops::Rem;
+    let p = &Bls12381FpParams::modulus();
+    let val = BigInt::from(value).rem(p);
+    bigint_to_fpelem(&val).unwrap()
+}
+
+impl<F> From<&AllocatedFieldElement<F>> for BlsFp
+where
+    F: PrimeField + PrimeFieldBits,
+{
+    fn from(value: &AllocatedFieldElement<F>) -> Self {
+        emulated_to_native(&value.0)
+    }
+}
+
+impl<F: PrimeField + PrimeFieldBits> AllocatedFieldElement<F> {
```

Comment:
The namespace names are pretty variable, with a few different styles/conventions being used. You might want to pick one and be consistent — maybe erring on the side of terse-but-descriptive.

Reply:
Something that might not be directly clear, most of the function names were direct translations of `gnark`'s functions (such as `MulBy014` becoming `mul_by_014` or something like that), so some things ended up with awkward names. I can do another pass to try to standardize the struct/function names a bit more.

Reply:
Oh, I misunderstood completely, thanks for the clarification. I agree that the lower level operations have inconsistent namespaces and I'll do a second pass changing them up, removing the `compute`s and simplifying them.
On the "higher-level" functions in other files I went with the pattern of e.g. `var <- a - b` for denoting specific assignments/operations corresponding to the variables in the code, and I usually went with the `x.double()`/`x.square()` pattern for denoting function calls, and the strings are generally more uniform there. (The main downside being that I ended up with basically a different namespace per operation and it makes it hard to read/parse the actual operations in the code because of all the `&mut cs.namespace("...")` pollution in the middle of the function calls/arguments)

Reply:
I've tried changing some things around but it was mostly a band-aid change to be honest. Let me know if there are other specific namespaces that still feel weirdly out of place or non-standardized and I'll change them around.

[rant]
In my ideal world, I wouldn't have defined any explicit namespace strings and a macro of sorts would take care of setting up the proper debugging/line/file/function information automagically for me. Having explicit namespaces should in theory be helpful for debugging, but in practice (for this gadget) when there is an error in some function, the unsatisfied constraint usually ends up being the equality constraint that compares the expected and calculated values, and you don't actually get any help from all the duplicated work of annotating individual operations with unique namespaces. The fact that all constraints must have a unique namespace path also means that I can't just try to ignore namespaces completely and just create a namespace-per-function and use that directly, since if I call `add` twice with the same namespace it will complain, so I would still need some kind of filler anyway.

I wouldn't mind writing all the namespaces manually if we had some nice debugging/profiling tools that could help analyze the circuit/constraint/aux variable space and identify/group/filter/etc which functions/namespaces are responsible for the majority of constraints and what call stack was responsible for generating the most constraints. Some kind of UI where it displays a flamegraph-like representation of the synthesized circuit. Unfortunately that doesn't exist (yet?), so it's just a pipe dream for now.
[/rant]